README.md
## tss

Annotate stdin with timestamps per line.
A Go port of [moreutils](https://joeyh.name/code/moreutils/)/ts and fork of [kevinburke/tss].

[kevinburke/tss]: https://github.com/kevinburke/tss

`tss` is like `ts` from moreutils,
but prints relative durations (with millisecond precision) by default,
and can be shipped as a compiled binary.

Try it out:

```console
$ (sleep 1; echo "hello"; sleep 2; echo "two sec") | tss
   1.00s    1.00s hello
   3.01s    2.01s two sec
```

The first column is the amount of time that has elapsed since the program started.
The second column is the amount of time that has elapsed since the last line printed.

![tss-demo](https://user-images.githubusercontent.com/30958501/103408835-923c9e80-4ba7-11eb-840a-7275e6fac18c.gif)

Note that: The demo has been created with the old docker version. The log format of `docker build` is different from the latest version.

[![license](https://img.shields.io/github/license/peaceiris/tss.svg)](https://github.com/peaceiris/tss/blob/main/LICENSE)
[![release](https://img.shields.io/github/release/peaceiris/tss.svg)](https://github.com/peaceiris/tss/releases/latest)
[![GitHub release date](https://img.shields.io/github/release-date/peaceiris/tss.svg)](https://github.com/peaceiris/tss/releases)
[![Release Feed](https://img.shields.io/badge/release-feed-yellow)](https://github.com/peaceiris/tss/releases.atom)
![Test](https://github.com/peaceiris/tss/workflows/CI/badge.svg?branch=main&event=push)
![Code Scanning](https://github.com/peaceiris/tss/workflows/Code%20Scanning/badge.svg?event=push)

[![Go Report Card](https://goreportcard.com/badge/github.com/peaceiris/tss)](https://goreportcard.com/report/github.com/peaceiris/tss)
[![CodeFactor](https://www.codefactor.io/repository/github/peaceiris/tss/badge)](https://www.codefactor.io/repository/github/peaceiris/tss)
[![Maintainability](https://api.codeclimate.com/v1/badges/5eaad3d1e44d6eb87a95/maintainability)](https://codeclimate.com/github/peaceiris/tss/maintainability)
[![codecov](https://codecov.io/gh/peaceiris/tss/branch/main/graph/badge.svg?token=4119ASAR7K)](https://codecov.io/gh/peaceiris/tss)



## Installation

### Homebrew

For macOS and Linux.

```sh
brew install peaceiris/tap/tss
```

[peaceiris/homebrew-tap/tss.rb](https://github.com/peaceiris/homebrew-tap/blob/main/Formula/tss.rb)

### Binary

[Releases · peaceiris/tss](https://github.com/peaceiris/tss/releases)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.6.1](https://github.com/peaceiris/tss/compare/v0.6.0...v0.6.1) (2021-12-15)


### ci

* bump actions/checkout from 2.3.4 to 2.3.5 (#56) ([84111ff](https://github.com/peaceiris/tss/commit/84111ff790a539e1ea2c6551375c50ff29fb1c37)), closes [#56](https://github.com/peaceiris/tss/issues/56)
* bump actions/checkout from 2.3.5 to 2.4.0 (#57) ([3702e22](https://github.com/peaceiris/tss/commit/3702e2215cce02c7664c7f9427bb52cc37f34afb)), closes [#57](https://github.com/peaceiris/tss/issues/57)
* bump codecov/codecov-action from 2.0.3 to 2.1.0 (#54) ([4b322df](https://github.com/peaceiris/tss/commit/4b322df4f72cf069c77f8992b11654003bb73321)), closes [#54](https://github.com/peaceiris/tss/issues/54)
* bump peaceiris/workflows from 0.10.0 to 0.11.0 (#55) ([fe0d5f9](https://github.com/peaceiris/tss/commit/fe0d5f9c29a16e5e06a14cf63539b195fed98d6a)), closes [#55](https://github.com/peaceiris/tss/issues/55)
* bump peaceiris/workflows from 0.9.0 to 0.10.0 (#53) ([708716e](https://github.com/peaceiris/tss/commit/708716e4233504cba894ba22c04f8907338cfd58)), closes [#53](https://github.com/peaceiris/tss/issues/53)

### deps

* bump github.com/spf13/cobra from 1.2.1 to 1.3.0 (#58) ([8bee5f9](https://github.com/peaceiris/tss/commit/8bee5f913ff9b658662a23d06fbbc661c33ef304)), closes [#58](https://github.com/peaceiris/tss/issues/58)

### test

* shuffle on ([d8b86dd](https://github.com/peaceiris/tss/commit/d8b86dd52a94ad69a08eb898825c336ac01caf9e))



# [0.6.0](https://github.com/peaceiris/tss/compare/v0.5.0...v0.6.0) (2021-08-28)


### build

* Add --update-cache (#45) ([31fe670](https://github.com/peaceiris/tss/commit/31fe670b1132cbb712071fac09e77aac58ca712c)), closes [#45](https://github.com/peaceiris/tss/issues/45)
* Add flagEnvNpm for npm_config_yes ([05e26ee](https://github.com/peaceiris/tss/commit/05e26eeb2ef56a82b5bb050ad23d50c4431d20ce))

### ci

* bump actions/cache from 2.1.5 to 2.1.6 (#39) ([3dd5c2b](https://github.com/peaceiris/tss/commit/3dd5c2be3dbefed1b7506f293494605489294f74)), closes [#39](https://github.com/peaceiris/tss/issues/39)
* bump actions/checkout from 2 to 2.3.4 (#36) ([43317d8](https://github.com/peaceiris/tss/commit/43317d8c35d0f4eb806805bc15209b534bd0d14c)), closes [#36](https://github.com/peaceiris/tss/issues/36)
* bump codecov/codecov-action from 1.5.0 to 1.5.1 (#40) ([66357ec](https://github.com/peaceiris/tss/commit/66357ec1d06cd1db0652a8682163af0a03ad745e)), closes [#40](https://github.com/peaceiris/tss/issues/40)
* bump codecov/codecov-action from 1.5.1 to 1.5.2 (#41) ([d86635f](https://github.com/peaceiris/tss/commit/d86635f076d787cc21aeded88ea45388605a31ee)), closes [#41](https://github.com/peaceiris/tss/issues/41)
* bump codecov/codecov-action from 1.5.2 to 2.0.1 (#46) ([c9dd504](https://github.com/peaceiris/tss/commit/c9dd504e6fb4cf7a44fa3c84b9f3259bd953a2db)), closes [#46](https://github.com/peaceiris/tss/issues/46)
* bump codecov/codecov-action from 2.0.1 to 2.0.2 (#47) ([48a39be](https://github.com/peaceiris/tss/commit/48a39be64d459381e822c7e5b51ae8ffed6cd796)), closes [#47](https://github.com/peaceiris/tss/issues/47)
* bump codecov/codecov-action from 2.0.2 to 2.0.3 (#49) ([7faccb2](https://github.com/peaceiris/tss/commit/7faccb273dc7c22a9c79c3dbe71d22db6899999e)), closes [#49](https://github.com/peaceiris/tss/issues/49)
* bump codecov/codecov-action from v1.4.0 to v1.4.1 (#33) ([d32afdd](https://github.com/peaceiris/tss/commit/d32afdd053d2ebd16a1609cc25066fe9308644ef)), closes [#33](https://github.com/peaceiris/tss/issues/33)
* bump codecov/codecov-action from v1.4.1 to v1.5.0 (#35) ([8037ea0](https://github.com/peaceiris/tss/commit/8037ea008e56a2cb300b9123984cbbc46efda428)), closes [#35](https://github.com/peaceiris/tss/issues/35)
* Use peaceiris/workflows/setup-go (#51) ([414af8a](https://github.com/peaceiris/tss/commit/414af8af48bfae162363a9c155e290a55c2f2ded)), closes [#51](https://github.com/peaceiris/tss/issues/51)

### deps

* bump github.com/spf13/cobra from 1.1.3 to 1.2.0 (#43) ([015d55d](https://github.com/peaceiris/tss/commit/015d55d9db7bfe240b16cc8e357105556253ce9a)), closes [#43](https://github.com/peaceiris/tss/issues/43)
* bump github.com/spf13/cobra from 1.2.0 to 1.2.1 (#44) ([09fa033](https://github.com/peaceiris/tss/commit/09fa0336cd03ac3915b977cd1a99027344a6ee87)), closes [#44](https://github.com/peaceiris/tss/issues/44)
* Go 1.17 (#52) ([9aa874c](https://github.com/peaceiris/tss/commit/9aa874c36e9a253d42ba342b70f97f0330f6a52c)), closes [#52](https://github.com/peaceiris/tss/issues/52)

### docs

* Add notes about demo ([7c8045e](https://github.com/peaceiris/tss/commit/7c8045ef15f7eb187e8f8ae8195a872f6f38d1d0)), closes [#18](https://github.com/peaceiris/tss/issues/18)



# [0.5.0](https://github.com/peaceiris/tss/compare/v0.4.0...v0.5.0) (2021-04-20)


### chore

* gofmt -w tools.go ([275f778](https://github.com/peaceiris/tss/commit/275f7788241d54f50052e1b78b36d97a1d3cfe9b))
* update description ([3c87a9e](https://github.com/peaceiris/tss/commit/3c87a9e9b4e54e0724c1dc87bc7135d5259f2d08))

### ci

* Add tools.go (#31) ([bf10f06](https://github.com/peaceiris/tss/commit/bf10f06837e44d349162d0a14fea63f530f689e6)), closes [#31](https://github.com/peaceiris/tss/issues/31)
* bump actions/cache from v2 to v2.1.4 (#23) ([6d958d0](https://github.com/peaceiris/tss/commit/6d958d05be15b4e7129124e1619dd2126429e3a5)), closes [#23](https://github.com/peaceiris/tss/issues/23)
* bump actions/cache from v2.1.4 to v2.1.5 (#29) ([6509162](https://github.com/peaceiris/tss/commit/6509162e411bd07f9ffdfda77cd546412abdf0af)), closes [#29](https://github.com/peaceiris/tss/issues/29)

main_test.go
package main

import (
	"testing"
)

func TestBuildEnvString(t *testing.T) {
	want := `TSS_BUILD_VERSION="0.1.2-development"
TSS_BUILD_COMMIT="aaeb6725631dcff02055855ee263ef5f45ed1eea-development"
TSS_BUILD_DATE="2020-12-28T11:01:32Z-development"
TSS_BUILD_GOOS="darwin"
TSS_BUILD_GOARCH="amd64"
TSS_BUILD_GOVERSION="go1.15.6"`
	out := BuildEnvString("darwin", "amd64", "go1.15.6")
	if out != want {
		t.Errorf("BuildEnvString(); want \n%s", want)
	}
}


cmd/tss_test.go
package tss_test

import (
	"bytes"
	"io"
	"strings"
	"testing"
	"time"

	tss "github.com/peaceiris/tss/cmd"
)

type sleepReader struct {
	count    int
	max      int
	sleepFor time.Duration
}

func (s *sleepReader) Read(p []byte) (int, error) {
	s.count++
	if s.count > s.max {
		return 0, io.EOF
	}
	if s.count == 1 {
		copy(p[:6], "hello\n")
		return 6, nil
	}
	time.Sleep(s.sleepFor)
	if s.count == 2 {
		copy(p[:3], "hel")
		return 3, nil
	}
	if s.count == 3 {
		copy(p[:3], "lo\n")
		return 3, nil
	}
	if s.count == 4 {
		copy(p[:15], "hello\nhello\nhel")
		return 15, nil
	}
	if s.count == 5 {
		copy(p[:3], "lo\n")
		return 3, nil
	}
	copy(p[:6], "hello\n")
	return 6, nil
}

func TestWriter(t *testing.T) {
	t.Parallel()
	max := 6
	s := &sleepReader{max: max, sleepFor: 2 * time.Millisecond}
	buf := new(bytes.Buffer)
	w := tss.NewWriter(buf, time.Time{})
	n, err := io.Copy(w, s)
	if err != nil {
		t.Fatal(err)
	}
	if int(n) != len("hello\n")*max {
		t.Errorf("expected n of 36, got %d:\n%s", n, buf)
	}
}

func TestCopy(t *testing.T) {
	t.Parallel()
	max := 6
	s := &sleepReader{max: max, sleepFor: 2 * time.Millisecond}
	buf := new(bytes.Buffer)
	n, err := tss.Copy(buf, s)
	want := len("hello\n") * 6
	if err != nil {
		t.Errorf("expected nil error, got %v", err)
	}
	if int(n) != want {
		t.Errorf("expected n of %d, got %d", want, n)
	}
	parts := strings.Split(buf.String(), "\n")
	if len(parts) != 7 {
		t.Errorf("incorrect number of parts: want 6 got %d:\n%q", len(parts), parts)
	}
	line1 := parts[0]
	if len(line1) != 23 {
		t.Errorf("line1 length: want %d got %d", 23, len(line1))
	}
	lineParts := strings.Fields(line1)
	if len(lineParts) != 2 {
		t.Errorf("wrong line parts")
	}
	part, err := time.ParseDuration(lineParts[0])
	if err != nil {
		t.Fatal(err)
	}
	if part > 100*time.Millisecond {
		t.Errorf("part took too long: %d", part)
	}
	lineParts = strings.Fields(parts[1])
	if len(lineParts) != 3 {
		t.Errorf("wrong number of line parts in line 2: got %d", len(lineParts))
	}
}

var scalerTests = []struct {
	in  time.Duration
	out string
}{
	{100 * time.Microsecond, "0.1ms"},
	{500 * time.Microsecond, "0.5ms"},
	{99 * time.Microsecond, "0.1ms"},
	{49 * time.Microsecond, "49.0µs"},
	{time.Millisecond, "1.0ms"},
	{56*time.Millisecond + 290*time.Microsecond, "56.3ms"},
	{56*time.Millisecond + 251*time.Microsecond, "56.3ms"},
	{56*time.Millisecond + 100*time.Microsecond, "56.1ms"},
	{3*time.Minute + 4*time.Second + 100*time.Millisecond, "3m04s"},
	{3*time.Minute + 14*time.Second + 100*time.Millisecond, "3m14s"},
	{14*time.Second + 100*time.Millisecond, "14.10s"},
	{2*time.Minute - 1*time.Microsecond, "2m00s"},
	{0, "0.0ms"},
}

func TestTimeScaler(t *testing.T) {
	for _, tt := range scalerTests {
		v := tss.TimeScaler(tt.in)
		if v != tt.out {
			t.Errorf("timeScaler(%q): want %q, got %q", tt.in, tt.out, v)
		}
	}
}


lib/tss_test.go
package tss_test

import (
	"bytes"
	"io"
	"strings"
	"testing"
	"time"

	tss "github.com/peaceiris/tss/lib"
)

type sleepReader struct {
	count    int
	max      int
	sleepFor time.Duration
}

func (s *sleepReader) Read(p []byte) (int, error) {
	s.count++
	if s.count > s.max {
		return 0, io.EOF
	}
	if s.count == 1 {
		copy(p[:6], "hello\n")
		return 6, nil
	}
	time.Sleep(s.sleepFor)
	if s.count == 2 {
		copy(p[:3], "hel")
		return 3, nil
	}
	if s.count == 3 {
		copy(p[:3], "lo\n")
		return 3, nil
	}
	if s.count == 4 {
		copy(p[:15], "hello\nhello\nhel")
		return 15, nil
	}
	if s.count == 5 {
		copy(p[:3], "lo\n")
		return 3, nil
	}
	copy(p[:6], "hello\n")
	return 6, nil
}

func TestWriter(t *testing.T) {
	t.Parallel()
	max := 6
	s := &sleepReader{max: max, sleepFor: 2 * time.Millisecond}
	buf := new(bytes.Buffer)
	w := tss.NewWriter(buf, time.Time{})
	n, err := io.Copy(w, s)
	if err != nil {
		t.Fatal(err)
	}
	if int(n) != len("hello\n")*max {
		t.Errorf("expected n of 36, got %d:\n%s", n, buf)
	}
}

func TestCopy(t *testing.T) {
	t.Parallel()
	max := 6
	s := &sleepReader{max: max, sleepFor: 2 * time.Millisecond}
	buf := new(bytes.Buffer)
	n, err := tss.Copy(buf, s)
	want := len("hello\n") * 6
	if err != nil {
		t.Errorf("expected nil error, got %v", err)
	}
	if int(n) != want {
		t.Errorf("expected n of %d, got %d", want, n)
	}
	parts := strings.Split(buf.String(), "\n")
	if len(parts) != 7 {
		t.Errorf("incorrect number of parts: want 6 got %d:\n%q", len(parts), parts)
	}
	line1 := parts[0]
	if len(line1) != 23 {
		t.Errorf("line1 length: want %d got %d", 23, len(line1))
	}
	lineParts := strings.Fields(line1)
	if len(lineParts) != 2 {
		t.Errorf("wrong line parts")
	}
	part, err := time.ParseDuration(lineParts[0])
	if err != nil {
		t.Fatal(err)
	}
	if part > 100*time.Millisecond {
		t.Errorf("part took too long: %d", part)
	}
	lineParts = strings.Fields(parts[1])
	if len(lineParts) != 3 {
		t.Errorf("wrong number of line parts in line 2: got %d", len(lineParts))
	}
}

var scalerTests = []struct {
	in  time.Duration
	out string
}{
	{100 * time.Microsecond, "0.1ms"},
	{500 * time.Microsecond, "0.5ms"},
	{99 * time.Microsecond, "0.1ms"},
	{49 * time.Microsecond, "49.0µs"},
	{time.Millisecond, "1.0ms"},
	{56*time.Millisecond + 290*time.Microsecond, "56.3ms"},
	{56*time.Millisecond + 251*time.Microsecond, "56.3ms"},
	{56*time.Millisecond + 100*time.Microsecond, "56.1ms"},
	{3*time.Minute + 4*time.Second + 100*time.Millisecond, "3m04s"},
	{3*time.Minute + 14*time.Second + 100*time.Millisecond, "3m14s"},
	{14*time.Second + 100*time.Millisecond, "14.10s"},
	{2*time.Minute - 1*time.Microsecond, "2m00s"},
	{0, "0.0ms"},
}

func TestTimeScaler(t *testing.T) {
	for _, tt := range scalerTests {
		v := tss.TimeScaler(tt.in)
		if v != tt.out {
			t.Errorf("timeScaler(%q): want %q, got %q", tt.in, tt.out, v)
		}
	}
}


lib/tss.go
package tss

import (
	"bytes"
	"fmt"
	"io"
	"strconv"
	"time"
)

// Writer represents output of tss
type Writer struct {
	w         io.Writer
	start     time.Time
	lastLine  time.Time
	buf       bytes.Buffer
	endOfLine bool
}

var oldestStart time.Time

func init() {
	// This is as early as we can get into measuring when the binary starts.
	oldestStart = time.Now()
}

// NewWriter creates a new Writer. If start is the zero value, the program's
// start time will be used.
func NewWriter(w io.Writer, start time.Time) *Writer {
	if start.IsZero() {
		start = oldestStart
	}
	return &Writer{w: w, start: start, endOfLine: true}
}

var padding = bytes.Repeat([]byte{' '}, 9)

// Write writes the contents of p into the buffer. It returns the number of
// bytes written. If nn < len(p), it also returns an error explaining why the
// write is short.
func (w *Writer) Write(p []byte) (int, error) {
	if len(p) == 0 {
		return 0, nil
	}
	w.buf.Reset()
	w.buf.Grow(len(p) + bytes.Count(p, []byte{'\n'})*len(padding))
	wrote := 0
	now := time.Now()
	pos := 0
	for {
		// write everything up to the next newline
		if w.endOfLine {
			// print timing info
			sinceStart := now.Sub(w.start).Round(100 * time.Microsecond)
			s := TimeScaler(sinceStart)
			for i := 0; i < 8-len(s); i++ {
				w.buf.WriteByte(' ')
			}
			w.buf.WriteString(s)
			w.buf.WriteByte(' ')
			if w.lastLine.IsZero() {
				w.buf.Write(padding)
				w.lastLine = now
			} else {
				sinceLastLine := now.Sub(w.lastLine).Round(100 * time.Microsecond)
				s := TimeScaler(sinceLastLine)
				for i := 0; i < 8-len(s); i++ {
					w.buf.WriteByte(' ')
				}
				w.buf.WriteString(s)
				w.buf.WriteByte(' ')
				w.lastLine = now
			}
			w.endOfLine = false
		}
		idx := bytes.IndexByte(p[pos:], '\n')
		if idx >= 0 {
			w.buf.Write(p[pos : pos+idx+1])
			wrote += idx + 1
			pos += idx + 1
			w.endOfLine = true
			if pos >= len(p) {
				break
			}
		} else {
			w.buf.Write(p[pos:])
			wrote += len(p) - pos
			break
		}
	}
	_, err := w.w.Write(w.buf.Bytes())
	return wrote, err
}

var forceNonZeroTestVal = time.Duration(0)

// TimeScaler returns a format string for the given Duration where all of the
// decimals will line up in the same column (fourth from the end).
func TimeScaler(d time.Duration) string {
	if d == 0 && forceNonZeroTestVal != 0 {
		d = forceNonZeroTestVal
	}
	switch {
	case d == 0:
		return "0.0ms"
	case d >= time.Minute:
		mins := (d + 30*time.Second) / time.Minute
		d = d - mins*time.Minute
		if d < 0 {
			d = 0 // can happen due to rounding
		}
		s := strconv.FormatFloat(float64(d.Nanoseconds())/1e9, 'f', 0, 64)
		return strconv.Itoa(int(mins)) + "m" + fmt.Sprintf("%02s", s) + "s"
	case d >= time.Second:
		return strconv.FormatFloat(float64(d.Nanoseconds())/1e9, 'f', 2, 64) + "s"
	case d >= 50*time.Microsecond:
		return strconv.FormatFloat(float64(d.Nanoseconds())/1e9*1000, 'f', 1, 64) + "ms"
	case d >= time.Microsecond:
		return strconv.FormatFloat(float64(d.Nanoseconds())/1e9*1000*1000, 'f', 1, 64) + "µs"
	default:
		return strconv.FormatFloat(float64(d.Nanoseconds()), 'f', 1, 64) + "ns"
	}
}

// Copy returns CopyTime
func Copy(w io.Writer, r io.Reader) (written int64, err error) {
	return CopyTime(w, r, time.Now())
}

// CopyTime copies from Reader to Writer until either EOF is reached on src or an error occurs.
func CopyTime(w io.Writer, r io.Reader, start time.Time) (written int64, err error) {
	return io.Copy(NewWriter(w, start), r)
}


cmd/version.go
package cmd

import (
	"fmt"
	"runtime"

	"github.com/spf13/cobra"
)

// tss build env
var (
	buildVersion string = "0.1.2-development"
	buildCommit  string = "aaeb6725631dcff02055855ee263ef5f45ed1eea-development"
	buildDate    string = "2020-12-28T11:01:32Z-development"
)

// NewCmdVersion creates version command
func NewCmdVersion() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "version",
		Short: "Print the version number of tss",
		Run: func(cmd *cobra.Command, args []string) {
			cmd.Printf(GetBuildEnvsString(runtime.GOOS, runtime.GOARCH, runtime.Version()))
		},
	}
	return cmd
}

// BuildEnv has shell environment name and value
type BuildEnv struct {
	name  string
	value string
}

// GetBuildEnvsString returns tss build environment variables
func GetBuildEnvsString(goOS string, goArch string, goVersion string) string {
	version := buildVersion
	commit := buildCommit

	buildEnvs := [6]BuildEnv{
		{name: "VERSION", value: version},
		{name: "COMMIT", value: commit},
		{name: "DATE", value: buildDate},
		{name: "GOOS", value: goOS},
		{name: "GOARCH", value: goArch},
		{name: "GOVERSION", value: goVersion},
	}

	out := ""
	for _, e := range buildEnvs {
		out += fmt.Sprintf("TSS_BUILD_%s=\"%s\"\n", e.name, e.value)
	}
	return out
}


magefile.go
//go:build mage
// +build mage

package main

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
	"time"

	"github.com/magefile/mage/mg"
	"github.com/magefile/mage/sh"
)

const (
	packageName = "github.com/peaceiris/tss"
)

var ldflags = `-X $PACKAGE/cmd.buildVersion=$VERSION -X $PACKAGE/cmd.buildCommit=$COMMIT_HASH -X $PACKAGE/cmd.buildDate=$BUILD_DATE`

// allow user to override go executable by running as GOEXE=xxx make ... on unix-like systems
var goexe = "go"

func init() {
	if exe := os.Getenv("GOEXE"); exe != "" {
		goexe = exe
	}

	// We want to use Go 1.11 modules even if the source lives inside GOPATH.
	// The default is "auto".
	os.Setenv("GO111MODULE", "on")
}

func runWith(env map[string]string, cmd string, inArgs ...interface{}) error {
	s := argsToStrings(inArgs...)
	return sh.RunWith(env, cmd, s...)
}

// Install development tools
func Setup() error {
	if err := sh.Run(goexe, "install", "golang.org/x/lint/golint"); err != nil {
		return err
	}
	return nil
}

// Build binary
func Build() error {
	return runWith(flagEnv(), goexe, "build", "-ldflags", ldflags, buildFlags(), packageName)
}

// Build binary with race detector enabled
func BuildRace() error {
	return runWith(flagEnv(), goexe, "build", "-race", "-ldflags", ldflags, buildFlags(), packageName)
}

// Install binary
func Install() error {
	return runWith(flagEnv(), goexe, "install", "-ldflags", ldflags, buildFlags(), packageName)
}

// Uninstall binary
func Uninstall() error {
	return sh.Run(goexe, "clean", "-i", packageName)
}

func flagEnv() map[string]string {
	hash, _ := sh.Output("git", "rev-parse", "--short", "HEAD")
	version, _ := sh.Output("git", "describe", "--tags")
	return map[string]string{
		"VERSION":     version,
		"PACKAGE":     packageName,
		"COMMIT_HASH": hash,
		"BUILD_DATE":  time.Now().Format("2006-01-02T15:04:05Z0700"),
	}
}

// Run tests and linters
func Check() {

	mg.Deps(Fmt, Vet)

	// don't run two tests in parallel, they saturate the CPUs anyway, and running two
	// causes memory issues in CI.
	mg.Deps(TestRace)
}

func testGoFlags() string {
	if isCI() {
		return ""
	}

	return "-test.short"
}

// Run tests
func Test() error {
	env := map[string]string{"GOFLAGS": testGoFlags()}
	return runCmd(env, goexe, "test", "-shuffle", "on", "-coverpkg", "./...", "-covermode", "atomic", "-coverprofile", "coverage.txt", "./...", buildFlags())
}

// Run tests with race detector
func TestRace() error {
	env := map[string]string{"GOFLAGS": testGoFlags()}
	return runCmd(env, goexe, "test", "-shuffle", "on", "-race", "-coverpkg", "./...", "-covermode", "atomic", "-coverprofile", "coverage.txt", "./...", buildFlags())
}

// Run gofmt -w
func Fmtw() error {
	if !isGoLatest() {
		return nil
	}
	pkgs, err := tssPackages()
	if err != nil {
		return err
	}
	for _, pkg := range pkgs {
		files, err := filepath.Glob(filepath.Join(pkg, "*.go"))
		if err != nil {
			return nil
		}
		for _, f := range files {
			_, err := sh.Output("gofmt", "-w", f)
			if err != nil {
				fmt.Printf("ERROR: running gofmt on %q: %v\n", f, err)
			}
		}
	}
	return nil
}

// Run gofmt linter
func Fmt() error {
	if !isGoLatest() {
		return nil
	}
	pkgs, err := tssPackages()
	if err != nil {
		return err
	}
	failed := false
	first := true
	for _, pkg := range pkgs {
		files, err := filepath.Glob(filepath.Join(pkg, "*.go"))
		if err != nil {
			return nil
		}
		for _, f := range files {
			// gofmt doesn't exit with non-zero when it finds unformatted code
			// so we have to explicitly look for output, and if we find any, we
			// should fail this target.
			s, err := sh.Output("gofmt", "-l", f)
			if err != nil {
				fmt.Printf("ERROR: running gofmt on %q: %v\n", f, err)
				failed = true
			}
			if s != "" {
				if first {
					fmt.Println("The following files are not gofmt'ed:")
					first = false
				}
				failed = true
				fmt.Println(f)
			}
		}
	}
	if failed {
		return errors.New("improperly formatted go files")
	}
	return nil
}

var (
	pkgPrefixLen = len("github.com/peaceiris/tss")
	pkgs         []string
	pkgsInit     sync.Once
)

func tssPackages() ([]string, error) {
	var err error
	pkgsInit.Do(func() {
		var s string
		s, err = sh.Output(goexe, "list", "./...")
		if err != nil {
			return
		}
		pkgs = strings.Split(s, "\n")
		for i := range pkgs {
			pkgs[i] = "." + pkgs[i][pkgPrefixLen:]
		}
	})
	return pkgs, err
}

// Run golint linter
func Lint() error {
	pkgs, err := tssPackages()
	if err != nil {
		return err
	}
	failed := false
	for _, pkg := range pkgs {
		// We don't actually want to fail this target if we find golint errors,
		// so we don't pass -set_exit_status, but we still print out any failures.
		if _, err := sh.Exec(nil, os.Stderr, nil, "golint", pkg); err != nil {
			fmt.Printf("ERROR: running go lint on %q: %v\n", pkg, err)
			failed = true
		}
	}
	if failed {
		return errors.New("errors running golint")
	}
	return nil
}

// Run go vet linter
func Vet() error {
	if err := sh.Run(goexe, "vet", "./..."); err != nil {
		return fmt.Errorf("error running go vet: %v", err)
	}
	return nil
}

func runCmd(env map[string]string, cmd string, args ...interface{}) error {
	if mg.Verbose() {
		return runWith(env, cmd, args...)
	}
	output, err := sh.OutputWith(env, cmd, argsToStrings(args...)...)
	if err != nil {
		fmt.Fprint(os.Stderr, output)
	}

	return err
}

func isGoLatest() bool {
	return strings.Contains(runtime.Version(), "1.17")
}

func isCI() bool {
	return os.Getenv("CI") != ""
}

func buildFlags() []string {
	if runtime.GOOS == "windows" {
		return []string{"-buildmode", "exe"}
	}
	return nil
}

func argsToStrings(v ...interface{}) []string {
	var args []string
	for _, arg := range v {
		switch v := arg.(type) {
		case string:
			if v != "" {
				args = append(args, v)
			}
		case []string:
			if v != nil {
				args = append(args, v...)
			}
		default:
			panic("invalid type")
		}
	}

	return args
}

func flagEnvNpm() map[string]string {
	return map[string]string{
		"npm_config_yes": "true",
	}
}

func BumpVersion(releaseType string) error {
	fmt.Printf("\nnpm_config_yes=true npx standard-version --release-as %s --preset eslint\n", releaseType)
	out, err := sh.OutputWith(flagEnvNpm(), "npx", "standard-version", "--release-as", releaseType, "--preset", "eslint")
	fmt.Println(out)
	if err != nil {
		return err
	}
	if err := sh.Run("git", "push", "origin", "main"); err != nil {
		return err
	}
	if err := sh.Run("git", "push", "origin", "--tags"); err != nil {
		return err
	}
	return nil
}

// Run npm_config_yes=true npx standard-version --release-as patch --preset eslint
func BumpPatchVersion() error {
	return BumpVersion("patch")
}

// Run npm_config_yes=true npx standard-version --release-as minor --preset eslint
func BumpMinorVersion() error {
	return BumpVersion("minor")
}

// Run npm_config_yes=true npx standard-version --release-as patch --preset eslint --dry-run
func BumpVersionTest() error {
	fmt.Printf("\nnpx standard-version --release-as patch --preset eslint --dry-run\n")
	out, err := sh.OutputWith(flagEnvNpm(), "npx", "standard-version", "--release-as", "patch", "--preset", "eslint", "--dry-run")
	fmt.Println(out)
	return err
}

// Run goreleaser check
func GoreleaserCheck() error {
	fmt.Printf("\ngoreleaser check\n")
	return sh.Run("goreleaser", "check")
}

// Run goreleaser --snapshot --skip-publish --rm-dist
func GoreleaserTest() error {
	fmt.Printf("\ngoreleaser --snapshot --skip-publish --rm-dist\n")
	return sh.Run("goreleaser", "--snapshot", "--skip-publish", "--rm-dist")
}

// Run GoreleaserCheck, GoreleaserTest, BumpVersionTest
func ReleaseTest() {
	mg.SerialDeps(GoreleaserCheck, GoreleaserTest, BumpVersionTest)
}
